### PR TITLE
Speed up frozen string dictionaries and sets by ~25-30%

### PR DESCRIPTION
**Pull request by geeknoid**
- Replace the specialized comparer types with specialized dictionary and set types, to avoid virtual dispatch overhead and enable better inlining.

### CODE REVIEW

File: src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs [hunk 1]
```diff
@@ -126,69 +101,37 @@ public static StringComparerBase Pick(ReadOnlySpan<string> uniqueStrings, bool i
...
-
-        private sealed class ComparerWrapper : IEqualityComparer<string>
-        {
-            private readonly SubstringComparerBase _comp;
-
-            public ComparerWrapper(SubstringComparerBase comp) => _comp = comp;
-
-            public bool Equals(string? x, string? y) => _comp.EqualsPartial(x, y);
-            public int GetHashCode([DisallowNull] string obj) => _comp.GetHashCode(obj);
+            results = new(true, ignoreCase, 0, 0, 0, 0);
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:134]: Same feedback about all of these we could consolidate into a single results setting using a bool for the first arg.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs [hunk 2]
```diff
@@ -238,5 +181,49 @@ private static double GetUniquenessFactor(HashSet<string> set, ReadOnlySpan<stri
 
             return set.Count / (double)uniqueStrings.Length;
         }
+
+        internal record struct AnalysisResults(
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs:185]: Please avoid using records.  They come with a whole bunch of extra gook we don't need.

File: src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/OrdinalStringFrozenDictionary.cs [hunk 3]
```diff
@@ -28,52 +32,46 @@ internal sealed class OrdinalStringFrozenDictionary<TValue> : FrozenDictionary<s
...
         private protected override TValue[] ValuesCore => _values;
-
-        /// <inheritdoc />
         private protected override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
-
-        /// <inheritdoc />
         private protected override int CountCore => _hashTable.Count;
 
-        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/OrdinalStringFrozenDictionary.cs:62]: It'd be great to add a comment about the pattern being used here.  It'd be really easy for someone to come along and see the overrides that just delegate to the base and think that's silly and remove it.

**Review by stephentoub — APPROVED** [review]

**Comment by geeknoid** [comment]
All comments addressed. Checking in once tests are done.

**Pull request merged**